[GStreamer] Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in EME

philn opened the pull request:
#### 7ce7b22758c84647943f33ac94e52559ca5f2b94
<pre>
[GStreamer] Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in EME
<a href="https://bugs.webkit.org/show_bug.cgi?id=285608">https://bugs.webkit.org/show_bug.cgi?id=285608</a>

Reviewed by Alicia Boya Garcia.

* Source/WebCore/platform/graphics/gstreamer/eme/CDMThunder.cpp:
(WebCore::CDMInstanceSessionThunder::CDMInstanceSessionThunder):
* Source/WebCore/platform/graphics/gstreamer/eme/GStreamerEMEUtilities.cpp:
(WebCore::markupStartElement):
(WebCore::markupEndElement):
(WebCore::markupText):
* Source/WebCore/platform/graphics/gstreamer/eme/GStreamerEMEUtilities.h:
(WebCore::ProtectionSystemEvents::ProtectionSystemEvents):
* Source/WebCore/platform/graphics/gstreamer/eme/WebKitThunderDecryptorGStreamer.cpp:

Canonical link: <a href="https://commits.webkit.org/288663@main">https://commits.webkit.org/288663@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3114dc6f6d92e33cdd8ea562048fa63868ab0a6e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84039 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3657 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38340 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89114 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35047 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86124 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3748 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11625 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65358 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23205 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87085 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2794 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76361 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45651 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2725 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30585 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34096 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73685 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31344 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90494 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11304 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8191 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/73815 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11528 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72190 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/73029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18058 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17325 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2651 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11256 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16728 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11104 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14580 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12876 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebCore/platform/graphics/gstreamer/eme/GStreamerEMEUtilities.h:
```diff
@@ -87,13 +88,9 @@ class ProtectionSystemEvents {
-            for (unsigned i = 0; streamEncryptionAllowedSystems[i]; ++i)
-                m_availableSystems.append(String::fromLatin1(streamEncryptionAllowedSystems[i]));
-        }
+        auto streamEncryptionAllowedSystems = span(static_cast<char**>(g_value_get_boxed(streamEncryptionAllowedSystemsValue)));
```
I understand this would be the overload that `span` would take here:

```
static inline std::span<char*> span(char** strv)
{
    auto size = g_strv_length(strv);
    return unsafeMakeSpan(strv, size);
}
```

However, according to [g_strv_length](https://docs.gtk.org/glib/func.strv_length.html), `str_array` should not be null and the previous code accounted for a possibly null value out of [g_value_get_boxed](https://docs.gtk.org/gobject/method.Value.get_boxed.html) (which the documentation states as a possibility). 

Do you think this is safe?

philn replied:
Should be, but a nullptr check wouldn't hurt :)

spenap commented on Source/WebCore/platform/graphics/gstreamer/eme/WebKitThunderDecryptorGStreamer.cpp:
```diff
@@ -84,21 +82,21 @@ static GRefPtr<GstCaps> createSinkPadTemplateCaps()
     }
     for (const auto& keySystem : supportedKeySystems) {
-        for (int i = 0; cencEncryptionMediaTypes[i]; ++i) {
+        for (auto& mediaType : cencEncryptionMediaTypes) {
```
could this be 

```suggestion
        for (const auto& mediaType : cencEncryptionMediaTypes) {
```
or does the structure need a non-const pointer?

philn replied:
const pointers work indeed, good catch!

spenap commented on Source/WebCore/platform/graphics/gstreamer/eme/WebKitThunderDecryptorGStreamer.cpp:
```diff
@@ -84,21 +82,21 @@ static GRefPtr<GstCaps> createSinkPadTemplateCaps()
 
     if (supportedKeySystems.contains(GStreamerEMEUtilities::s_WidevineKeySystem) || supportedKeySystems.contains(GStreamerEMEUtilities::s_ClearKeyKeySystem)) {
-        for (int i = 0; webmEncryptionMediaTypes[i]; ++i) {
+        for (auto& mediaType :  webmEncryptionMediaTypes) {
```
```suggestion
        for (const auto& mediaType :  webmEncryptionMediaTypes) {
```

ditto

csaavedra commented on Source/WebCore/platform/graphics/gstreamer/eme/GStreamerEMEUtilities.h:
```diff
@@ -87,13 +88,12 @@ class ProtectionSystemEvents {
+            return;
+        auto streamEncryptionAllowedSystems = span(static_cast<char**>(systemsArray));
+        for (unsigned i = 0; i < streamEncryptionAllowedSystems.size(); ++i)
+            m_availableSystems.append(String::fromLatin1(streamEncryptionAllowedSystems[i]));
```
I think you could use here something like:

```c++
    for (const auto system : span(static_cast<char**>(systemsArray)))
        m_availableSystems.append(String::fromLatin1(system));
```

or so.

csaavedra reviewed (commented):
Just a nit.

philn committed:
[GStreamer] Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in EME
https://bugs.webkit.org/show_bug.cgi?id=285608

Reviewed by Alicia Boya Garcia.

* Source/WebCore/platform/graphics/gstreamer/eme/CDMThunder.cpp:
(WebCore::CDMInstanceSessionThunder::CDMInstanceSessionThunder):
* Source/WebCore/platform/graphics/gstreamer/eme/GStreamerEMEUtilities.cpp:
(WebCore::markupStartElement):
(WebCore::markupEndElement):
(WebCore::markupText):
* Source/WebCore/platform/graphics/gstreamer/eme/GStreamerEMEUtilities.h:
(WebCore::ProtectionSystemEvents::ProtectionSystemEvents):
* Source/WebCore/platform/graphics/gstreamer/eme/WebKitThunderDecryptorGStreamer.cpp:

Canonical link: https://commits.webkit.org/288663@main
